src/mgr/Mgr.cc

```diff
@@ -496,6 +503,42 @@ void Mgr::handle_service_map(MServiceMap *m)
   server.got_service_map();
 }
 
+void Mgr::handle_mon_map(MMonMap *m)
+{
+  dout(20) << __func__ << dendl;
+  assert(lock.is_locked_by_me());
+  // read monmap from monclient instead of decoding from *m.
+  monc->with_monmap([this](const MonMap &m) { monmap = m; });
+  m->put();
```

Comment:
no need to save the monmap here. i think we just need to cull the removed monitors from the `daemon_state`, like:

```c++
  std::set<std::string> names_exist;
  cluster_state.with_monmap([&] (auto &monmap) {
    for (unsigned int i = 0; i < monmap.size(); i++) {
      names_exist.insert(monmap.get_name(i));
    }
  });
  daemon_state.cull("mon", names_exist);
```

and let `DaemonServer::handle_report()` to take care of the new joiners.